src/core/server/saved_objects/migrationsv2/actions/check_for_unknown_docs.ts

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import * as Either from 'fp-ts/lib/Either';
+import * as TaskEither from 'fp-ts/lib/TaskEither';
+import { estypes } from '@elastic/elasticsearch';
+import type { SavedObjectsRawDocSource } from '../../serialization';
+import { ElasticsearchClient } from '../../../elasticsearch';
+import {
+  catchRetryableEsClientErrors,
+  RetryableEsClientError,
+} from './catch_retryable_es_client_errors';
+
+/** @internal */
+export interface CheckForUnknownDocsParams {
+  client: ElasticsearchClient;
+  indexName: string;
+  unusedTypesQuery: estypes.QueryDslQueryContainer;
+  knownTypes: string[];
+}
+
+/** @internal */
+export interface CheckForUnknownDocsResponseDoc {
+  id: string;
+  type: string;
+}
+
+/** @internal */
+export interface CheckForUnknownDocsResponse {
+  unknownDocs: CheckForUnknownDocsResponseDoc[];
+}
+
+export const checkForUnknownDocs = ({
+  client,
+  indexName,
+  unusedTypesQuery,
+  knownTypes,
+}: CheckForUnknownDocsParams): TaskEither.TaskEither<
+  RetryableEsClientError,
+  CheckForUnknownDocsResponse
+> => () => {
+  const query = createUnknownDocQuery(unusedTypesQuery, knownTypes);
+
+  return client
+    .search<SavedObjectsRawDocSource>({
+      index: indexName,
+      body: {
+        query,
+      },
+    })
+    .then((response) => {
+      const { hits } = response.body.hits;
+      return Either.right({
+        unknownDocs: hits.map((hit) => ({ id: hit._id, type: hit._source?.type ?? 'undefined' })),
+      });
+    })
+    .catch(catchRetryableEsClientErrors);
+};
+
+const createUnknownDocQuery = (
+  unusedTypesQuery: estypes.QueryDslQueryContainer,
+  knownTypes: string[]
+): estypes.QueryDslQueryContainer => {
+  return {
+    bool: {
+      must: unusedTypesQuery,
+      must_not: knownTypes.map((type) => ({
+        term: {
+          type,
+        },
+      })),
+    },
+  };
```

Comment:
I included the `unusedTypesQuery` in the unknown doc queries, are we're not migrating them to the temp index, and as it would allow to 'remove' old types by unregistering them while adding them to the unused query.

Reply:
Does it mean that users *must* remove them to continue the migration? If so, it might be considered as a breaking change even though we just fixed a bug.

Reply:
No, I mean the opposite actually: I'm excluding the objects matching the unusedTypesQuery, to avoid failing if any is encountered given than we don't migrate them anyway.

Reply:
ok, `must` usage is a bit misleading here because `unusedTypesQuery` contains another `must_not` inside.

Reply:
Yea, the whole `unusedTypesQuery` naming is misleading because it's actually an `excludeUnusedTypesQuery`